### ISSUE TITLE
Phase 3 — Integration & Ship: publish @mailwoman/neural@0.1.0 to npm

### ISSUE THREAD

**Issue by GirlBossRush** [isp-nexus/mailwoman#11]
# Phase 3 — TypeScript Integration & Ship

**Goal:** integrate the trained model into Mailwoman via `NeuralSequenceClassifier`, wire it through the policy system, publish `@mailwoman/neural@0.1.0` and weight packages to npm. End state: a user can `npm install @mailwoman/neural @mailwoman/neural-weights-en-us` and get neural classification for coarse components.

**Duration estimate:** 1.5 weeks.

**Branch:** `neural/phase-3-integration`

**Depends on:** Phase 2 complete with weights packages on disk.

## Pre-flight

- [ ] Phase 2 weights packages exist and pass eval
- [ ] You have npm publish access (or know how to coordinate with the maintainer for publishing)
- [ ] `onnxruntime-node` installs cleanly in the dev environment

### 1. Package scaffolding

- [ ] Create `packages/neural/` workspace
- [ ] Dependencies: `onnxruntime-node`, SentencePiece WASM (pick one — `@bloomberg/sentencepiece` or a maintained alternative; verify and document in `DECISIONS.md`), `@mailwoman/core`
- [ ] Strict TypeScript config

### 2. SentencePiece tokenizer wrapper

- [ ] `packages/neural/src/tokenizer.ts`
- [ ] Loads the SentencePiece model from a weights package
- [ ] Exposes `encode(text: string): { ids: number[]; tokens: string[]; offsets: [number, number][] }`
- [ ] The `offsets` field is critical — it's how we map model output BIO labels back to character spans in the original input.

⚠ **Tokenizer parity check** — the single most important test in this phase:

- [ ] `packages/neural/test/tokenizer-parity.test.ts`
- [ ] Loads 10k `(raw, tokens)` pairs from corpus-v0.1.0
- [ ] Re-tokenizes each `raw` with the TS SentencePiece
- [ ] Asserts byte-for-byte equality with the stored `tokens`
- [ ] If a single example fails, the integration is broken. Fix before continuing.

### 3. ONNX inference wrapper

- [ ] `packages/neural/src/onnx-runner.ts`
- [ ] Loads ONNX model via `onnxruntime-node`
- [ ] Exposes `infer(tokenIds: number[]): Promise<number[][]>` returning logits per token
- [ ] Handles batching (group multiple sections for one inference call)
- [ ] Configurable execution provider (CPU default, CUDA via env var)
- [ ] Lazy loading: model loads on first `infer` call, not on construction (unless `warmup: true`)

### 4. BIO decoding

- [ ] `packages/neural/src/decode.ts`
- [ ] Input: logits per token, original token offsets
- [ ] Output: list of `(span, component, confidence)` tuples
- [ ] BIO decoding rules:
  - `B-T` starts a new span
  - `I-T` continues the previous span if it was `B-T` or `I-T`, else treat as `B-T` (graceful)
  - `O` ends any open span
- [ ] Confidence per span = mean softmax probability of the predicted label across the span's tokens
- [ ] Test cases: well-formed BIO, ill-formed BIO (recovery), all-O input, single-token spans

### 5. `NeuralSequenceClassifier`

- [ ] `packages/neural/src/sequence-classifier.ts`
- [ ] Implements `Classifier` from `@mailwoman/core`
- [ ] Constructor takes `NeuralSequenceClassifierOptions`, resolves and loads the weights package
- [ ] `classify(section, context)`:
  1. Tokenize the section's text
  2. Run ONNX inference
  3. Decode BIO labels to spans
  4. Filter by `minConfidence`
  5. Map spans back to `Span` objects in the original input (use SentencePiece offsets, then translate to original-input offsets via section's start offset)
  6. Emit `ClassificationProposal[]` with `source: 'neural'`, `source_id` from the model card version
- [ ] Returns empty array if model fails to load (graceful degradation per #7)

### 6. Weights package loader

- [ ] `packages/neural/src/weights.ts`
- [ ] `loadWeights(packageName: string): Promise<{ modelBytes, tokenizerBytes, modelCard }>`
- [ ] Resolves package via Node module resolution
- [ ] Reads files from package's distribution directory
- [ ] Caches in memory across instances of `NeuralSequenceClassifier`

### 7. Solver integration

- [ ] In `packages/core/src/solver.ts` (or wherever the existing solver entry lives), add the neural classifier path:
  - If a `LocaleProfile` has a `weightsPackage`, instantiate a `NeuralSequenceClassifier` for that locale
  - Run it alongside rule classifiers during the classify phase
  - All proposals flow through `PolicyRegistry.apply` before solving
- [ ] No solver logic changes needed — the abstraction from Phase 0 was designed for this.

### 8. Policy migration: `country` → `neural_preferred`

The first Ship of Theseus step.

- [ ] Update `packages/core/src/policy-defaults.ts`:
  - `country`: from `rule_only` to `neural_preferred` with `confidence_threshold: 0.8`
- [ ] Run the full test suite + golden set eval
- [ ] Verify on the golden set that `country` accuracy improved (or held) and no other component regressed
- [ ] If anything regresses, revert the policy change and investigate. The migration is gated on metrics, not on intent.

▶ Repeat the same for `region` if Phase 2 metrics justify. Coarse components only. Do not migrate `locality` or below in Phase 3 — defer until Stage 2 trains.

### 9. End-to-end test

- [ ] `packages/neural/test/e2e.test.ts`
- [ ] Test: `parse("123 Main St, Portland, OR 97215", { locale: 'en-US' })`
  - Neural classifier emits `country: USA` (inferred even though "USA" is absent from input)
  - Rule classifier still emits `postcode: 97215`
  - Final solution has both
  - `source` fields are correctly populated
- [ ] Test: same with `locale: 'fr-FR'` and a French address
- [ ] Test: missing weights package → graceful degradation, rule-only solution

### 10. CLI surfacing

- [ ] `npx mailwoman parse --locale en-US --neural ...` opts into neural classifier
- [ ] `npx mailwoman parse --locale en-US --no-neural ...` opts out (rule-only)
- [ ] Default: neural on if weights package is installed, off otherwise
- [ ] Debug output: when neural is on, show which classifier produced each component in the output

### 11. Documentation

- [ ] `packages/neural/README.md` — installation, usage, configuration, the `LocaleProfile` extension point
- [ ] Update root #15 with a new "Neural" section
- [ ] Migration guide: how an existing Mailwoman user opts into the neural classifier
- [ ] Performance numbers from the benchmark suite

### 12. Performance & benchmarks

- [ ] `packages/neural/bench/` with vitest-bench or simple scripts
- [ ] Benchmark: cold load time, warm classification latency p50/p95/p99, throughput, memory footprint
- [ ] All within budgets in #7
- [ ] If neural is slower than budgeted: profile. Likely culprits: ONNX session not reused across calls, batching not enabled, SentencePiece WASM not warm.

### 13. Publish

- [ ] Final version bumps:
  - `@mailwoman/core` → minor bump (added types)
  - `@mailwoman/classifiers` → minor bump (adapter additions)
  - `@mailwoman/neural@0.1.0` → new, marked beta
  - `@mailwoman/neural-weights-en-us@0.1.0` → new, marked beta
  - `@mailwoman/neural-weights-fr-fr@0.1.0` → new, marked beta
- [ ] Check licenses on every package's `package.json` — AGPL-3.0 for code, weights packages need their own license decision (see Phase 6 license note)
- [ ] `npm publish` each. Use `--access public` for scoped packages.
- [ ] Verify install in a clean directory: `npm install @mailwoman/neural @mailwoman/neural-weights-en-us` and run a sample parse.

⚠ Before publishing weights, settle the license question. OSM is ODbL (share-alike). WOF is CC0. BAN is Licence Ouverte. A model trained on ODbL data is arguably ODbL — talk to a lawyer or pick a clear license posture and document it.

## Success criteria checklist

- [ ] Tokenizer parity test green on 10k examples
- [ ] End-to-end test green for en-US and fr-FR
- [ ] Benchmarks within budget
- [ ] Golden set: country F1 improved (or held), nothing regressed
- [ ] `npx mailwoman parse` works with `--locale` and uses neural classifier when available
- [ ] Packages published to npm
- [ ] Documentation updated
- [ ] Branch tagged `neural-phase-3-complete`
- [ ] Blog post draft (optional but recommended) in `docs/blog/2026-...-neural-classifier.md`

## When the human will want a check-in

After Phase 3 ships, pause. The human (the project creator) should look at:

- Real-world feedback from a couple of test deployments
- The blog post draft
- License decisions on weights packages
- Whether to proceed to Stage 2 (street-level) or pause to gather usage data

Do not begin Phase 4 (geocoder fusion) or even Stage 2 of Phase 1/2 (street-level training) without explicit confirmation. The point of shipping is to learn.

## When to call this phase done

When `npm install @mailwoman/neural @mailwoman/neural-weights-en-us` followed by a `parse()` call works on a clean machine and produces output with `source: 'neural'` proposals for the `country` component.

---
Part of epic #15.

**Comment by GirlBossRush** [comment]
## Output-shape design clarification (operator framing, 2026-05-18)

Address parsing is inherently lossy — you can't extract the original formatted address from the parsed result. The closest you can get is by applying a formatter to the parsed result. Three legitimate output shapes worth supporting:

| Shape | Preserves | Loses | Round-trip fidelity |
|---|---|---|---|
| **JSON object** `{locality: "Paris", postcode: "75005"}` | tag → value | order, repetition, hierarchy, untagged tokens | Country-specific formatter required; reorder errors silent. libpostal/Pelias compatible. |
| **Tuple array** `[["locality","Paris"], ["postcode","75005"]]` | tag → value + **original order** + repetition | hierarchy | Trivially faithful to ordering; reconstruction is `array.map(([_,v]) => v).join(" ")`. |
| **S-expression** `(locality "Paris" (postcode "75005"))` | order + **hierarchy** + arbitrary attributes (`:conf`, `:src`) | nothing material | Most expressive; natural carrier for Phase 4 source annotations. |

**Operator's theory** (worth testing in this issue): S-expressions encode hierarchy and may best preserve the original input shape for re-formatting. The flat JSON form is what existing libpostal/Pelias-style consumers expect; the S-expr form is what users who want lossless round-trip should ask for.

## Scope addition for this issue

1. Implement all three decoders (`decodeAsJson`, `decodeAsTuples`, `decodeAsSexp`) consuming the same BIO output from the neural model
2. Add a CLI flag: `--format json|tuple|sexp`; default = `json` for libpostal-compat
3. Add a **round-trip fidelity eval**: `parse(format(parse(raw))) == parse(raw)` over the golden set, reported per-format. If S-expr wins decisively, flip default
4. Port the existing isp-nexus formatter (`isp-nexus/universe/mailwoman/postal/formatting.ts`) as the JSON-decoder's inverse path; build a separate `sexp→string` formatter for the S-expr path

**No model retraining required.** This is pure post-processing on the same BIO output the Phase 2 model already emits.

## Related Phase 4 angle

If/when the Phase 4 Resolver lands, S-expressions become the natural carrier for source annotations:

```
(locality "Paris" :src wof-admin :conf 0.97
  (postcode "75005" :src wof-postalcode :conf 0.93))
```

JSON-object has no clean way to express `:src` / `:conf` per-component without resorting to nested metadata dicts; tuple arrays can sort-of-fake it but it gets ugly. S-expr puts attributes where they belong. So Phase 3's output-shape work + Phase 4's Resolver work compound: by the time #12 ships, the format is ready to carry the annotations.

Related: #12 (Phase 4 Resolver), #43 (v0.2.0 retrain — the model whose output these decoders consume).

**Comment by GirlBossRush** [comment]
## Output-shape decoder — pinning the XML spec

Filed as PR #58 ahead of the full `@mailwoman/neural` package scaffolding, since the decoder is pure logic with no inference deps.

### Three projections from one `AddressTree`

- **JSON** — libpostal-compat `tag → value`, first-occurrence wins.
- **Tuples** — `Array<[tag, value]>`, source-ordered, repetition-safe.
- **XML** — nested mixed-content with attributes (chosen over S-expression for LLM-tooling alignment + off-the-shelf parsers).

### XML shape

```xml
<address raw="15 Rue de Rivoli, 75004 Paris, France">
  <country start="29" end="35" conf="0.98">France
    <locality start="22" end="27" conf="0.97">Paris
      <postcode start="16" end="21" conf="0.99">75004</postcode>
      <street start="3" end="15" conf="0.88">Rue de Rivoli
        <house_number start="0" end="2" conf="0.99">15</house_number>
      </street>
    </locality>
  </country>
</address>
```

- Element's direct text node = component's own value
- Children = sub-components by containment hierarchy
- `conf` attr always; `start`/`end` preserve source order alongside the containment-derived element order
- `src` attribute reserved for Phase 4 (Resolver source provenance), not emitted yet
- Sibling order is source order (`start` ascending)

### Containment rule

Each tag has a priority-ordered parent list (`PARENT_OF` in `packages/core/core/decoder/containment.ts`). Parent resolution picks the highest-priority tag that has a labeled span, with char-distance tiebreaker for repeated parents.

### ⚠ DOM gotcha

`element.textContent` on a mixed-content node returns concatenated descendant text. Documented in the file header; consumers use XPath `text()` or DOM child-node filtering.

### Phase 3 task delta

Original task list said "expose `decode_components`" (libpostal-style flat). That now becomes three exports — `decodeAsJson` / `decodeAsTuples` / `decodeAsXml` — all consuming the same `AddressTree`. CLI flag in the future `@mailwoman/neural` package will surface as `--format json|tuple|xml`, default `json`.

**Comment by GirlBossRush** [comment]
Shipped — @mailwoman/neural published to npm.
